src/recovery.rs

```diff
@@ -0,0 +1,48 @@
+use crate::{
+    member::UnitMessage, units::UncheckedSignedUnit, Data, Hasher, NodeCount, NodeIndex, Recipient,
+    Signature,
+};
+use futures::{
+    channel::mpsc::{UnboundedReceiver, UnboundedSender},
+    future::ready,
+    SinkExt, StreamExt,
+};
+
+const CATCH_UP_SECS: u64 = 5;
+
+// input:
+
+// Given network sink and stream, recovers all units created by us in this instance of consensus.
+pub(crate) async fn recover_newest_unit<H: Hasher, D: Data, S: Signature>(
+    node_count: NodeCount,
+    responses_newest: UnboundedReceiver<Option<UncheckedSignedUnit<H, D, S>>>,
+    mut messages_for_peers: UnboundedSender<(UnitMessage<H, D, S>, Recipient)>,
+    our_index: NodeIndex,
+) -> Option<UncheckedSignedUnit<H, D, S>> {
+    messages_for_peers
+        .send((UnitMessage::RequestNewest(our_index), Recipient::Everyone))
+        .await
+        .expect("send should succeed");
+    // receive at most N responses for CATCH_UP_SECS seconds, ignore the invalid ones, and
+    // choose the one with highest round
+    // (we assume that we also receive the request and respond to it)
+    responses_newest
+        .take(node_count.0)
```

Comment:
Unless we have signatures under the responses or can track the origin of network messages (which we cannot as of now), then relying on the number of received responses is not a great idea, because an adversary could just send N responses. I think the simplest for now would be to just wait CATCH_UP_SECS.

Reply:
The current solution relies on the assumption that there is no adversary, and even if we tracked the origin, we are in trouble without a salt, but I agree that this quick fix is the best we can do for now.